File: plane.py
from math import sqrt


class Plane:
    def __init__(self, point, normal, material):
        self.point = point
        self.normal = normal
        self.material = material
        
    def hit(self, r):
        t = ((self.point - r.origin).dot(self.normal)) / (r.direction.dot(self.normal))
        if t > 0:
            return t
        return None
    
    def normal_at(self, pos):
        return self.normal
    
    

File: shapes/plane.cpp
#include "plane.h"

Plane::Plane(const Vector &point, const Vector &normal,
             const std::shared_ptr<Material>& material)
    : Shape(material), point(point), normal(normal) {
}

double Plane::hit(const Ray &ray) const {
    double denom = ray.direction.dot(normal);
    if (denom > 1e-6) {
        Vector p0l0 = point - ray.origin;
        auto t = p0l0.dot(normal) / denom;
        return t;
    }
    return -1;
}

Vector Plane::normalAt(const Vector &p) const {
    return normal;
}